File: models/developing/cantgap/driver_beamgap1.lua
-- HiQLab
-- Copyright (c): Regents of the University of California

 
beamf = loadfile 'beamgap.lua'
beamf();
mesh:initialize();

numid = mesh:get_numid();
u  = QArray:new(numid, 1);
R  = QArray:new(numid, 1);
du = QArray:new(numid, 1);

-- FIXME: I need a way to move the no-mass directions up front.

for VV=10,60 do

  -- Set new voltage
  printf('Voltage %e', VV);
  V = VV;
  mesh:apply_bc()

  -- Modified Newton corrector
  K = mesh:assemble_dR(1,0,0);
  for k = 1,8 do
    mesh:assemble_R();
    mesh:get_f(R);
    K:solve(du,R);
    u:sub(du);
    mesh:set_u(u)
    if print_newton then
      printf('  %d: %e %e', k, du:normf(), R:normf());
    end
  end
  K:delete()

  -- Compute eigenvalues on reduced system
  K = mesh:assemble_dR(1,0,0);
  M = mesh:assemble_dR(0,0,1);

  d,v = leigs{
    op = function(x,y)
      scratch = M:apply(x,scratch)
      K:solve(y,scratch)
    end,
    n = numid,
    real = true,
    mode = 'standard',
    which = 'LM',
    nev = 1
  }
  printf('  Freq: %e kHz', sqrt(1/d(1)) / 2 / pi);

  K:delete()
  M:delete()
end

File: models/developing/cantgap/driver_beamgap1.m
% HiQLab
% Copyright (c): Regents of the University of California
% $Id: driver_beamgap1.m,v 1.1 2005/06/28 19:05:44 dbindel Exp $

function [Vall, fall] = driver_beamgap1;

print_newton = 1;
popt.deform = 1;
popt.cfields = 3;

[mesh,Ls] = Mesh_load('beamgap.lua'); 
numid = Mesh_get_numid(mesh);

u  = zeros(numid,1);

Vall = 10:60;
lall = zeros(1,length(Vall));

for iV=1:length(Vall)

  % Set new voltage
  VV = Vall(iV);
  fprintf('Voltage %e\n', VV);
  Lua_set_double(Ls, 'V', VV);
  Mesh_apply_bc(mesh);

  % Modified Newton corrector
  K = Mesh_assemble_dR(mesh, 1,0,0);
  for k = 1:8 
    R  = Mesh_assemble_R(mesh);
    du = K\R;
    u  = u-du;
    Mesh_set_u(mesh, u);
    if print_newton
      fprintf('  %d: %e %e\n', k, norm(du), norm(R));
    end
  end

  % Show the mesh
  plotfield2d(mesh, popt);
  shg;

  % Compute eigenvalues on reduced system
  [M,K] = Mesh_assemble_mk(mesh);
  [L,U,P,Q] = qlu(K);
  opts.disp = 0;
  d = eigs(@Kfun,numid, 3, 'LM', opts, L,U,P,Q, M);
  f = sqrt(1/d(1)) / 2 / pi;
  fall(iV) = f;
  fprintf('  Freq: %e kHz\n', f);

end

Mesh_delete(mesh);

% Plot voltage vs frequency
figure(2);
plot(Vall, fall);
xlabel('Potential (V)');
ylabel('Frequency (kHz)');


% -- Apply K\M

function [y] = Kfun(x, L, U, P, Q, M)

y = Q*(U\(L\(P*(M*x))));
